ballerina/fragment_visitor.bal

```diff
@@ -20,16 +20,26 @@ class FragmentVisitor {
     *parser:Visitor;
 
     private ErrorDetail[] errors;
-    private map<string> usedFragments;
+    private string[] usedFragments;
     private parser:DocumentNode documentNode;
+    private string[] visitedFragments;
 
     public isolated function init(parser:DocumentNode documentNode) {
         self.errors = [];
-        self.usedFragments = {};
+        self.usedFragments = [];
         self.documentNode = documentNode;
+        self.visitedFragments = [];
     }
 
     public isolated function validate() returns ErrorDetail[]? {
+        foreach parser:FragmentNode fragmentNode in self.documentNode.getFragments() {
+            if !fragmentNode.isInlineFragment() && self.visitedFragments.indexOf(fragmentNode.getName()) is () {
+                self.detectCycles(fragmentNode, {});
+            }
+        }
+        if self.errors.length() > 0 {
+            return self.errors;
+        }
         self.visitDocument(self.documentNode);
         if self.errors.length() > 0 {
             return self.errors;
```

Comment:
Is there a way to remove this duplicated code? I understand why the same check has to be done twice, but can we remove the duplication somehow?